src/commands/lock.rs

```diff
@@ -0,0 +1,122 @@
+}
+
+/// Represents the lock file structure.
+#[derive(Debug, Serialize, Deserialize)]
+struct Lock {
+    /// The time when the lock file was created.
+    #[serde(rename = "generation_time")]
+    timestamp: String,
+    /// A set of workflow names that were found in the source.
+    workflows: HashMap<String, String>,
+    /// A mapping of Docker image names to their sha256 digests.
+    images: HashMap<String, String>,
+}
+
+/// Performs the `lock` command.
+pub async fn lock(args: Args) -> Result<()> {
+    let s = args.source.unwrap_or(Source::Directory(PathBuf::from(
+        std::path::Component::CurDir.as_os_str(),
+    )));
+    let results = match Analysis::default().add_source(s).run().await {
+        Ok(results) => results,
+        Err(errors) => {
+            // SAFETY: this is a non-empty, so it must always have a first
+            // element.
+            bail!(errors.into_iter().next().unwrap())
+        }
+    };
+
+    let mut workflows = HashMap::new();
+    let mut images: HashSet<String> = HashSet::new();
+    let mut buffer = String::new();
+    for result in results {
+        let doc = result.document().root();
+
+        if let Some(wf) = result.document().workflow() {
+            workflows.insert(wf.name().to_owned(), result.document().uri().to_string());
+        }
+
+        for task in result.document().tasks() {
+            doc.ast()
+                .as_v1()
+                .expect("should be a v1 document")
+                .tasks()
+                .filter(|t| t.name().text() == task.name())
+                .for_each(|t| {
+                    if let Some(runtime) = t.runtime() {
+                        if let Some(container) = runtime.container() {
+                            if let Ok(image) = container.value() {
+                                if let Expr::Literal(LiteralExpr::String(s)) = image.expr() {
+                                    if let Some(text) = s.text() {
+                                        text.unescape_to(&mut buffer);
+                                        images.insert(buffer.clone());
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    buffer.clear();
+                });
+        }
+    }
+
+    let time = Utc::now();
+
+    let mut map: HashMap<String, String> = HashMap::new();
+    for image in images {
+        let prefix = image.split(':').next().unwrap_or("");
+        let docker = crankshaft_docker::with_defaults()?;
+
+        let i = docker
+            .inner()
+            .inspect_registry_image(&image, None)
+            .await
+            .expect("should inspect registry image");
+
+        // Insert the manifest digest into the map.
+        map.insert(
+            image.clone(),
+            prefix.to_owned() + "@" + &i.descriptor.digest.expect("should have a digest"),
+        );
+    }
+
+    if !map.is_empty() {
```

Comment:
Theoretically someone passes in a WDL `Source` that results in an empty map here. What's the desired UX? I think writing an empty lock is preferable to doing nothing.